File: mobile/lib/core/shoot_client_profiles.dart
import 'package:hsdash_mobile/core/calendar_utils.dart';
import 'package:hsdash_mobile/core/weddings_archive_index.dart';
import 'package:hsdash_mobile/models/shoot_calendar_entry.dart';
import 'package:hsdash_mobile/models/shoot_client_profile.dart';

String shootClientDisplayLabel({
  required String brideName,
  required String groomName,
  required String clientName,
}) {
  final b = brideName.trim();
  final g = groomName.trim();
  if (b.isNotEmpty && g.isNotEmpty) return '$b & $g';
  if (b.isNotEmpty) return b;
  if (g.isNotEmpty) return g;
  return clientName.trim();
}

ShootClientProfile shootClientProfileFromEntry(ShootCalendarEntry entry) {
  final bride = entry.brideName?.trim() ?? '';
  final groom = entry.groomName?.trim() ?? '';
  final clientName = entry.clientName.trim();
  final rawPhone = entry.phoneNumber?.trim() ?? '';
  final rawContact = entry.clientContact?.trim() ?? '';
  final phone = effectiveClientPhone(phoneNumber: rawPhone, clientContact: rawContact);
  final contact = effectiveClientContact(phoneNumber: rawPhone, clientContact: rawContact);
  final city = entry.city?.trim() ?? '';
  final venue = entry.venue?.trim() ?? '';
  final type = _meaningfulClientType(entry.clientType?.trim() ?? '');
  final isWedding = bride.isNotEmpty ||
      groom.isNotEmpty ||
      (entry.clientType?.toLowerCase().contains('wedding') ?? false) ||
      clientName.contains('&');

  final label = shootClientDisplayLabel(brideName: bride, groomName: groom, clientName: clientName);

  return ShootClientProfile(
    id: weddingKeyForEntry(entry),
    displayLabel: label.isNotEmpty ? label : clientName,
    clientName: clientName.isNotEmpty ? clientName : label,
    clientType: type,
    clientContact: contact,
    city: city,
    venue: venue,
    brideName: bride,
    groomName: groom,
    phoneNumber: phone,
    isWedding: isWedding,
  );
}

String _meaningfulClientType(String type) {
  final t = type.trim();
  if (t.isEmpty) return '';
  final lower = t.toLowerCase();
  if (lower == 'wedding' || lower == 'other') return '';
  return t;
}

ShootClientProfile _normalizeProfile(ShootClientProfile p) {
  final rawPhone = p.phoneNumber;
  final rawContact = p.clientContact;
  return ShootClientProfile(
    id: p.id,
    displayLabel: p.displayLabel,
    clientName: p.clientName,
    clientType: p.clientType,
    city: p.city,
    venue: p.venue,
    brideName: p.brideName,
    groomName: p.groomName,
    isWedding: p.isWedding,
    phoneNumber: effectiveClientPhone(phoneNumber: rawPhone, clientContact: rawContact),
    clientContact: effectiveClientContact(phoneNumber: rawPhone, clientContact: rawContact),
  );
}

bool _entryIsNewer(ShootCalendarEntry a, ShootCalendarEntry b) {
  return shootDayKey(a.day).compareTo(shootDayKey(b.day)) >= 0;
}

List<ShootClientProfile> shootClientProfilesFromEntries(
  Iterable<ShootCalendarEntry> entries, {
  String search = '',
}) {
  final byKey = <String, ShootCalendarEntry>{};

  for (final entry in entries) {
    final name = entry.clientName.trim();
    final bride = entry.brideName?.trim() ?? '';
    final groom = entry.groomName?.trim() ?? '';
    if (name.isEmpty && bride.isEmpty && groom.isEmpty) continue;

    final key = weddingKeyForEntry(entry);
    final prev = byKey[key];
    if (prev == null || _entryIsNewer(entry, prev)) {
      byKey[key] = entry;
    }
  }

  var list = byKey.values.map((e) => _normalizeProfile(shootClientProfileFromEntry(e))).toList()
    ..sort((a, b) => a.displayLabel.compareTo(b.displayLabel));

  return filterClientProfiles(list, search);
}

List<String> _profileSearchLabels(ShootClientProfile c) {
  final labels = <String>[
    normalizeClientLabel(c.displayLabel),
    normalizeClientLabel(c.clientName),
  ];
  if (c.brideName.isNotEmpty || c.groomName.isNotEmpty) {
    labels.add(
      normalizeClientLabel(
        resolveShootClientName(isWedding: true, brideName: c.brideName, groomName: c.groomName, clientName: ''),
      ),
    );
    labels.add(normalizeClientLabel('${c.brideName} ${c.groomName}'));
  }
  return labels.where((s) => s.isNotEmpty).toSet().toList();
}

/// Fuzzy match — handles "yesh and harika" vs stored "Yesh & Harika".
bool profileMatchesQuery(ShootClientProfile profile, String query) {
  final needle = normalizeClientLabel(query);
  if (needle.isEmpty) return false;

  for (final hay in _profileSearchLabels(profile)) {
    if (hay == needle || hay.contains(needle) || needle.contains(hay)) return true;
    final words = needle.split(' ').where((w) => w.length >= 2).toList();
    if (words.isNotEmpty && words.every((w) => hay.contains(w))) return true;
  }
  return false;
}

List<ShootClientProfile> filterClientProfiles(
  List<ShootClientProfile> profiles,
  String query, {
  int limit = 8,
}) {
  if (query.trim().isEmpty) return [];

  final matches = profiles.where((c) => profileMatchesQuery(c, query)).toList();

  return matches.length <= limit ? matches : matches.sublist(0, limit);
}

/// When the typed name clearly matches one saved client, return it for auto-prefill.
ShootClientProfile? resolveClientProfileForQuery(List<ShootClientProfile> profiles, String query) {
  final needle = normalizeClientLabel(query);
  if (needle.length < 3) return null;

  final matches = profiles.where((c) => profileMatchesQuery(c, query)).toList();
  if (matches.isEmpty) return null;

  for (final c in matches) {
    for (final label in _profileSearchLabels(c)) {
      if (label == needle) return c;
    }
  }

  if (matches.length == 1) {
    final only = matches.first;
    for (final label in _profileSearchLabels(only)) {
      if (label.startsWith(needle) && needle.length >= 6) return only;
    }
  }

  return null;
}

String _profileMergeKey(ShootClientProfile p) {
  final label = p.displayLabel.isNotEmpty ? p.displayLabel : p.clientName;
  final key = normalizeClientLabel(label);
  return key.isNotEmpty ? key : p.id;
}

ShootClientProfile _combineProfiles(ShootClientProfile base, ShootClientProfile extra) {
  return ShootClientProfile(
    id: base.id,
    displayLabel: base.displayLabel.isNotEmpty ? base.displayLabel : extra.displayLabel,
    clientName: base.clientName.isNotEmpty ? base.clientName : extra.clientName,
    clientType: base.clientType.isNotEmpty ? base.clientType : extra.clientType,
    clientContact: base.clientContact.isNotEmpty ? base.clientContact : extra.clientContact,
    city: base.city.isNotEmpty ? base.city : extra.city,
    venue: base.venue.isNotEmpty ? base.venue : extra.venue,
    brideName: base.brideName.isNotEmpty ? base.brideName : extra.brideName,
    groomName: base.groomName.isNotEmpty ? base.groomName : extra.groomName,
    phoneNumber: base.phoneNumber.isNotEmpty ? base.phoneNumber : extra.phoneNumber,
    isWedding: base.isWedding || extra.isWedding,
  );
}

/// Prefer [primary] (calendar rows with venue) when merging with API client list.
List<ShootClientProfile> mergeClientProfiles(
  List<ShootClientProfile> primary,
  List<ShootClientProfile> secondary,
) {
  final byKey = <String, ShootClientProfile>{};
  for (final p in secondary) {
    byKey[_profileMergeKey(p)] = p;
  }
  for (final p in primary) {
    final key = _profileMergeKey(p);
    final prev = byKey[key];
    byKey[key] = prev == null ? p : _combineProfiles(p, prev);
  }
  final out = byKey.values.toList()..sort((a, b) => a.displayLabel.compareTo(b.displayLabel));
  return out;
}

File: frontend/src/lib/shootClientProfiles.ts
import { weddingKeyForEntry, normalizeClientLabel } from "@/lib/weddingsArchiveIndex";
import { calendarDayKeyFromIso } from "@/lib/calendarUtils";
import {
  effectiveClientContact,
  effectiveClientPhone,
  resolveShootClientName,
  type ShootClientProfile,
} from "@/types/shootClientProfile";
import type { ShootCalendarEntry } from "@/types/domain";

function shootDayKey(day: string) {
  return calendarDayKeyFromIso(day);
}

export function shootClientDisplayLabel(opts: { brideName: string; groomName: string; clientName: string }) {
  const b = opts.brideName.trim();
  const g = opts.groomName.trim();
  if (b && g) return `${b} & ${g}`;
  if (b) return b;
  if (g) return g;
  return opts.clientName.trim();
}

function meaningfulClientType(type: string) {
  const t = type.trim();
  if (!t) return "";
  const lower = t.toLowerCase();
  if (lower === "wedding" || lower === "other") return "";
  return t;
}

export function shootClientProfileFromEntry(entry: ShootCalendarEntry): ShootClientProfile {
  const bride = entry.brideName?.trim() ?? "";
  const groom = entry.groomName?.trim() ?? "";
  const clientName = entry.clientName.trim();
  const rawPhone = entry.phoneNumber?.trim() ?? "";
  const rawContact = entry.clientContact?.trim() ?? "";
  const phone = effectiveClientPhone(rawPhone, rawContact);
  const contact = effectiveClientContact(rawPhone, rawContact);
  const city = entry.city?.trim() ?? "";
  const venue = entry.venue?.trim() ?? "";
  const type = meaningfulClientType(entry.clientType?.trim() ?? "");
  const isWedding =
    !!bride ||
    !!groom ||
    (entry.clientType?.toLowerCase().includes("wedding") ?? false) ||
    clientName.includes("&");

  const label = shootClientDisplayLabel({ brideName: bride, groomName: groom, clientName });

  return {
    id: weddingKeyForEntry(entry),
    displayLabel: label || clientName,
    clientName: clientName || label,
    clientType: type,
    clientContact: contact,
    city,
    venue,
    brideName: bride,
    groomName: groom,
    phoneNumber: phone,
    isWedding,
  };
}

function normalizeProfile(p: ShootClientProfile): ShootClientProfile {
  return {
    ...p,
    phoneNumber: effectiveClientPhone(p.phoneNumber, p.clientContact),
    clientContact: effectiveClientContact(p.phoneNumber, p.clientContact),
  };
}

function entryIsNewer(a: ShootCalendarEntry, b: ShootCalendarEntry) {
  return shootDayKey(a.day).localeCompare(shootDayKey(b.day)) >= 0;
}

export function shootClientProfilesFromEntries(entries: Iterable<ShootCalendarEntry>, search = "") {
  const byKey = new Map<string, ShootCalendarEntry>();
  for (const entry of entries) {
    const name = entry.clientName.trim();
    const bride = entry.brideName?.trim() ?? "";
    const groom = entry.groomName?.trim() ?? "";
    if (!name && !bride && !groom) continue;
    const key = weddingKeyForEntry(entry);
    const prev = byKey.get(key);
    if (!prev || entryIsNewer(entry, prev)) byKey.set(key, entry);
  }
  const list = [...byKey.values()]
    .map((e) => normalizeProfile(shootClientProfileFromEntry(e)))
    .sort((a, b) => a.displayLabel.localeCompare(b.displayLabel));
  return filterClientProfiles(list, search);
}

function profileSearchLabels(c: ShootClientProfile) {
  const labels = new Set([normalizeClientLabel(c.displayLabel), normalizeClientLabel(c.clientName)]);
  if (c.brideName || c.groomName) {
    labels.add(
      normalizeClientLabel(
        resolveShootClientName({ isWedding: true, brideName: c.brideName, groomName: c.groomName, clientName: "" }),
      ),
    );
    labels.add(normalizeClientLabel(`${c.brideName} ${c.groomName}`));
  }
  return [...labels].filter(Boolean);
}

export function profileMatchesQuery(profile: ShootClientProfile, query: string) {
  const needle = normalizeClientLabel(query);
  if (!needle) return false;
  for (const hay of profileSearchLabels(profile)) {
    if (hay === needle || hay.includes(needle) || needle.includes(hay)) return true;
    const words = needle.split(" ").filter((w) => w.length >= 2);
    if (words.length > 0 && words.every((w) => hay.includes(w))) return true;
  }
  return false;
}

export function filterClientProfiles(profiles: ShootClientProfile[], query: string, limit = 8) {
  if (!query.trim()) return [];
  const matches = profiles.filter((c) => profileMatchesQuery(c, query));
  return matches.length <= limit ? matches : matches.slice(0, limit);
}

export function resolveClientProfileForQuery(profiles: ShootClientProfile[], query: string) {
  const needle = normalizeClientLabel(query);
  if (needle.length < 3) return null;
  const matches = profiles.filter((c) => profileMatchesQuery(c, query));
  if (!matches.length) return null;
  for (const c of matches) {
    for (const label of profileSearchLabels(c)) {
      if (label === needle) return c;
    }
  }
  if (matches.length === 1) {
    const only = matches[0]!;
    for (const label of profileSearchLabels(only)) {
      if (label.startsWith(needle) && needle.length >= 6) return only;
    }
  }
  return null;
}

function profileMergeKey(p: ShootClientProfile) {
  const label = p.displayLabel || p.clientName;
  const key = normalizeClientLabel(label);
  return key || p.id;
}

function combineProfiles(base: ShootClientProfile, extra: ShootClientProfile): ShootClientProfile {
  return {
    id: base.id,
    displayLabel: base.displayLabel || extra.displayLabel,
    clientName: base.clientName || extra.clientName,
    clientType: base.clientType || extra.clientType,
    clientContact: base.clientContact || extra.clientContact,
    city: base.city || extra.city,
    venue: base.venue || extra.venue,
    brideName: base.brideName || extra.brideName,
    groomName: base.groomName || extra.groomName,
    phoneNumber: base.phoneNumber || extra.phoneNumber,
    isWedding: base.isWedding || extra.isWedding,
  };
}

export function mergeClientProfiles(primary: ShootClientProfile[], secondary: ShootClientProfile[]) {
  const byKey = new Map<string, ShootClientProfile>();
  for (const p of secondary) byKey.set(profileMergeKey(p), p);
  for (const p of primary) {
    const key = profileMergeKey(p);
    const prev = byKey.get(key);
    byKey.set(key, prev ? combineProfiles(p, prev) : p);
  }
  return [...byKey.values()].sort((a, b) => a.displayLabel.localeCompare(b.displayLabel));
}

export function profileFromApiJson(json: Record<string, unknown>): ShootClientProfile {
  const bride = String(json.brideName ?? "");
  const groom = String(json.groomName ?? "");
  const type = String(json.clientType ?? "");
  const clientName = String(json.clientName ?? "");
  const isWedding =
    !!bride.trim() ||
    !!groom.trim() ||
    type.toLowerCase().includes("wedding") ||
    clientName.includes("&");
  const rawPhone = String(json.phoneNumber ?? "");
  const rawContact = String(json.clientContact ?? "");
  return normalizeProfile({
    id: String(json.id ?? ""),
    displayLabel: String(json.displayLabel ?? clientName),
    clientName,
    clientType: type,
    clientContact: effectiveClientContact(rawPhone, rawContact),
    city: String(json.city ?? ""),
    venue: String(json.venue ?? ""),
    brideName: bride,
    groomName: groom,
    phoneNumber: effectiveClientPhone(rawPhone, rawContact),
    isWedding,
  });
}
